File: 14.py
class LinearProbingHash:
    def __init__(self, size):
        self.size = size
        self.table = [None] * size
        self.deleted = "<DEL>"

    def insert(self, value):
        idx = value % self.size
        start = idx
        while self.table[idx] not in (None, self.deleted):
            if self.table[idx] == value:
                return  # already exists
            idx = (idx + 1) % self.size
            if idx == start:
                return  # full
        self.table[idx] = value

    def search(self, value):
        idx = value % self.size
        start = idx
        while self.table[idx] is not None:
            if self.table[idx] == value:
                return idx
            idx = (idx + 1) % self.size
            if idx == start:
                break
        return -1

    def delete(self, value):
        idx = self.search(value)
        if idx != -1:
            self.table[idx] = self.deleted


# Driver Example
h = LinearProbingHash(10)
h.insert(10)
h.insert(20)
h.insert(17)
print(h.search(17))
h.delete(20)
print(h.search(20))

File: 14.js
class LinearProbingHash {
    constructor(size) {
        this.size = size;
        this.table = new Array(size).fill(null);
        this.DELETED = "<DEL>";
    }

    insert(value) {
        let idx = value % this.size;
        const start = idx;

        while (this.table[idx] !== null && this.table[idx] !== this.DELETED) {
            if (this.table[idx] === value) return;
            idx = (idx + 1) % this.size;
            if (idx === start) return; // full
        }
        this.table[idx] = value;
    }

    search(value) {
        let idx = value % this.size;
        const start = idx;

        while (this.table[idx] !== null) {
            if (this.table[idx] === value) return idx;
            idx = (idx + 1) % this.size;
            if (idx === start) break;
        }
        return -1;
    }

    delete(value) {
        const idx = this.search(value);
        if (idx !== -1) this.table[idx] = this.DELETED;
    }
}

// Example
const h = new LinearProbingHash(10);
h.insert(10);
h.insert(20);
h.insert(17);
console.log(h.search(17));
h.delete(20);
console.log(h.search(20));
